### PR TITLE
Implementation of climate platform

### PR DESCRIPTION
**Pull request by vmartinv**
This is yet another version of the implementation that was never finished in #72, inspired by @rigibe 's. Supersedes #538 

# Background
Tuya thermostats all have very different dp settings. If one is too general the config is too hard to understand, can't be properly validated and/or is hard to maintain. If it's too specific then it will constantly require updates to add compatibility for more devices. I've looked at the config and DPs other users have posted and I think the schema I created here achieves a good balance.

# Config design decisions
These are the issues and how they're being dealt:
- *auto vs Auto, on vs heating, etc*: I followed the same approach as in the cover device. The user is able to choose from a known list of options. I think I can manage adding the missing ones, if any. They are really easy to add.
- *ECO sometimes is in a dedicated DP*: The user specifies which dp the ECO is located, it can be the same as the one used for the HVAC mode.
- *automatic HVAC_MODE_OFF*: This mode is automatically used when the id (thermostat switch) is off. Setting a different mode also turns on the switch.

## Details
- For some reason my thermostat didn't respond when updating two dps in a row, so I had to add a 0.1s delay in this very specific case.
- My thermostat reports with 0.1 precision (230=23C), but the target temperature is set with 1 precision (23=23C), so I had to add an optional `target_precision`.

# Results
This leads to a long list of options (see screenshot of the complete config for my thermostat). But contrary to the alternatives (JSONs, extra forms with lists) they're really intuitive and I can fill it out easily. It's currently working for my thermostat. I plan to write a wiki about the options if this gets added.

Screenshot 2021-08-10 at 19 35 45

## Possible improvements

- I think we can remove `euristic_action` to reduce complexity, I think most thermostats report their state already.
- Any missing sets?

### CODE REVIEW

**Comment by josephus** [comment]
This is how I got it to work with my ME82 thermostat:

image
image

My DPs are the following with a few assumptions I made that could be useful to add more options on the UI:

> {
>   '1': true, // On/Off (?)
>   '2': 'manual', // auto or manual mode
>   '10': false, // Anti-freeze mode (?)
>   '16': 22, // Target temp
>   '19': 40,  // Max temp
>   '24': 225, // Current temp * 10
>   '27': -1, // Temp correction offset
>   '36': 'close', // Relay state
>   '43': 'out', // Sensor selection 'in' or 'out'
>   '45': 0, // Child lock(?)
>   '101': 1, // Factory reset(?)
>   '102': '52' // (?)
> }

The only two issue I found:

- The entity keeps dropping to unavailable state every few hours or so for a few seconds. Perhaps we should be more liberal with timeouts?
- What's the point of selecting device ID from the list of DPs? I actually don't know what to select here.

**Comment by larius-z** [comment]
I keep my fingers crossed for this project, and can’t wait to see it released! Great job :)

File: custom_components/localtuya/climate.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement used by the platform."""
+        if (
+            self._config.get(CONF_TEMPERATURE_UNIT, DEFAULT_TEMPERATURE_UNIT)
+            == TEMPERATURE_FAHRENHEIT
+        ):
+            return TEMP_FAHRENHEIT
+        return TEMP_CELSIUS
```
[the remark this answers is not in the document]
Comment (vmartinv) [reply custom_components/localtuya/climate.py:216]: I think the current approach is cleaner

**Comment by vmartinv** [comment]
@rospogrigio @postlund are you able to check this, the code has been tested in several devices, I fixed all suggestions and linters seem to be passing

**Comment by vmartinv** [comment]
@josephus the error you mentioned doesn't seem to be related to this implementation. Perhaps you can check your logs for any errors. The device ID field is used in the same way as in other devices, usually is the DP 1.

**Comment by vmartinv** [comment]
@jellespijker add suggestions added, can you approve if it still works for you?

**Comment by jellespijker** [comment]
@vmartinv thanks. I will check. I might have some time tomorrow to check the changes at the latest on Saturday.

**Comment by maslyankov** [comment]
Any updates on this?

**Comment by srodighiero** [comment]
@rospogrigio @postlund have you the possibility to review it? This would be a great feature integration, and would let lot of people (me included) to totally dismiss Tuya integration for Local Tuya only.

**Comment by rospogrigio** [comment]
I can review it but you have to resolve some conflicts due to recent merging of another PR, please do.

**Review by siminkuv — APPROVED** [review]

**Comment by vmartinv** [comment]
@rospogrigio conflicts resolved

**Comment by patrik-malina** [comment]
I can confirm it works to me

**Comment by rospogrigio** [comment]
Sorry @vmartinv I merged another PR and more conflicts arose, can you please resolve them?

**Comment by krzysiek2788** [comment]
I have tested this with my Thermostats and it is working pretty well. I have termostat something similar to BHT-002 for eletric heating mats below are DP's

```
1(value:True) -> on/off
2(value:manual) -> mode
3(value:Warming) -> state?
8(value:false) -> anti freezing?
10(value:false) -> open window sensor?
16(value:210) -> desired temp
19(value:450) -> max temp
24(value:205) -> current temp
25(value:close) -> Window open/close?
27(value:-2) -> air temp calibration
40(value:close) -> child lock
43(value:in) -> temp sensor in/out/all
101(value:10) -> histeresis? which should be vaule=value/10
102(value:500) -> max floor temp
```

File: custom_components/localtuya/climate.py [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+HVAC_MODE_SETS = {
+    "manual/auto": {
+        HVAC_MODE_HEAT: "manual",
+        HVAC_MODE_AUTO: "auto",
+    },
+    "Manual/Auto": {
+        HVAC_MODE_HEAT: "Manual",
+        HVAC_MODE_AUTO: "Auto",
+    },
```
Comment (Alloc186) [inline custom_components/localtuya/climate.py:67]: "Manual/Program": {
        HVAC_MODE_HEAT: "Manual",
        HVAC_MODE_AUTO: "Program",
    },

File: custom_components/localtuya/climate.py [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    "True/False": {
+        HVAC_MODE_HEAT: True,
+    },
+}
+HVAC_ACTION_SETS = {
+    "True/False": {
+        CURRENT_HVAC_HEAT: True,
+        CURRENT_HVAC_OFF: False,
```
Comment (Alloc186) [inline custom_components/localtuya/climate.py:75]: CURRENT_HVAC_IDLE: False,

**Review by Alloc186 — APPROVED** [review]

**Comment by krzysiek2788** [comment]
As I can see it isn't changing state on Climate card, it is always heating while in fact it is not heating, so it isn't switching to idle state. I tried to set different DP's in integration but it isn't working.
image
image
image

**Comment by larius-z** [comment]
> As I can see it isn't changing state on Climate card, it is always heating while in fact it is not heating, so it isn't switching to idle state. I tried to set different DP's in integration but it isn't working.
> image
> image
> image

Please double check attribute mapping. In my case I mixed the correct attribute with the one which represents on/off state of the entire thermostat.

**Comment by krzysiek2788** [comment]
@larius-z I tried that and nothing helps my dp's configuration looks like this:
image

**Comment by vmartinv** [comment]
@krzysiek2788 it’s not working because your device reports status as Warming (you selected heating/no_heating). Can you tell me the string that DP has when not warming so I can add it as an option?

**Comment by krzysiek2788** [comment]
@vmartinv I thought so. On my there is Heat and Warming. Heat = heating, Warming = not heating

**Comment by vmartinv** [comment]
Added feedback from @Alloc186 and @krzysiek2788 , please approve it after successful testing.

**Comment by krzysiek2788** [comment]
@vmartinv working but it shouldn't be Off but Inactive/idle like on the left termostat if possible:
image

**Comment by vmartinv** [comment]
@krzysiek2788 check the latest commit, I have set it to use idle

**Review by krzysiek2788 — APPROVED** [review]
I can confirm that for my tuya thermostats all settings are working very well and as should work. Also information about heating/idle are showing correct information

**Comment by Alloc186** [comment]
I think I found a bug, but it's more of a HA feature.

Try this commands in developer option or automation: climate.turn_on, climate.turn_off

climate.turn_off works OK (HVAC_MODE_OFF), it turns off the thermostat itself
but command climate.turn_on, sets thermostat to heat only as default mode (HVAC_MODE_HEAT).

Problem is, if you had set it to Auto before (HVAC_MODE_AUTO), after calling control service climate.turn_on, it changes mode to Manual. So it doesn't keep its previous state as expected.

These command should probably only set DP1 to True or False to turn device ON or OFF, what you think?

**Review by siminkuv — APPROVED** [review]

**Comment by vmartinv** [comment]
@Alloc186 you're right. This is because turn on/off method was not implemented, I added it now and the behaviour is now as you expect it. Accept it if it works for you.

**Comment by Alloc186** [comment]
> @Alloc186 you're right. This is because turn on/off method was not implemented, I added it now and the behaviour is now as you expect it. Accept it if it works for you.

WOW! Now it works superbly, thank you Sir. I owe you coffee ;)

**Comment by al3x1337** [comment]
Hi all, hoping to get some assistance with installing this for myself and possibly others who are looking to test this build out ahead of the official release. 
In my case I want to test it out to see if it will work with a Heat Storm HS-1000-WX space heater.

I've downloaded the .zip for the _climate_platform_ branch and pasted the localtuya folder into custom_components, overriding my existing files there installed by HACS. When restarting i get an error in the logs that localtuya can't be started and in Integrations it does not up anymore. 
Pasting the localtuya folder from the master branch or reinstalling via HACS fixes this.

On another fresh Hassio instance i've pasted the same localtuya folder into custom_components and did a restart. Nothing in the logs and does not show up within Add Integrations.

How do I install this climate branch?

**Comment by javimurcia** [comment]
The climate branch is an unrelated branch, if you check this pull request attributes, it is being made against main.

You can download the main branch, and get the files changed by this pull request, fro. If you the "files changed tab" and manually replace them.

Alternatively, if you are proficient with git cli, you can clone this repo on the command-line, and checkout this pull request.

**Review by Alloc186 — APPROVED** [review]

**Review by siminkuv — APPROVED** [review]

**Comment by vmartinv** [comment]
@rospogrigio all the issues were addressed, I thinks this is ready to be merged

**Review by rospogrigio — APPROVED** [review]

**Comment by yarosman** [comment]
Is it possible to use climate platform with air conditioner ?

**Comment by vmartinv** [comment]
> Is it possible to use climate platform with air conditioner ?

if it offers the similar interface yes, but nobody can answer it with this data. I'd suggest to try it and if it doesn't work create a separate issue with as much information as possible, such as dp's, model, functions, etc.

**Comment by yarosman** [comment]
> > Is it possible to use climate platform with air conditioner ?
> 
> if it offers the similar interface yes, but nobody can answer it with this data. I'd suggest to try it and if it doesn't work create a separate issue with as much information as possible, such as dp's, model, functions, etc.

at least they have different mode
"values": "{\"range\":[\"auto\",\"cold\",\"hot\",\"wet\",\"wind\"]}"
and other parameters.

**Comment by vmartinv** [comment]
> I'd suggest to try it and if it doesn't work create a separate issue with as much information as possible, such as dp's, model, functions, etc.

Again, I'd suggest to try it and if it doesn't work create a separate issue with as much information as possible, such as dp's, model, functions, etc. Thanks.

**Pull request merged**